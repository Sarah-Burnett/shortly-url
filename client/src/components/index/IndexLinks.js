import React, { useEffect } from "react";
import styled from "styled-components";
import CopyShortLink from "./CopyShortLink";
import LinkCopyButton from "./LinkCopyBtn";
import { breakpoint, grey, alignPadding } from "../Styles";
import { useCopyLink } from "../../hooks/useCopyLink";
import { useLinks } from "../../hooks/useLinks";
import { CSSTransition, TransitionGroup } from "react-transition-group";

const LinkContainer = styled.div`
	padding: 2vh ${alignPadding};
	height: min-content;
	margin: auto;
	background: ${grey};
`;

const Link = styled.div`
	min-height: 15vh;
	background: #fff;
	padding: 2vh 2vw;
	margin: 2vh 0;
	border-radius: 15px;
	@media (min-width: ${breakpoint}) {
		display: flex;
		align-items: center;
		height: 10vh;
	}
`;

const FullLink = styled.div`
	flex: 3;
	height: 5vh;
	line-height: 5vh;
	width: 100%;
	overflow-y: auto;
	overflow-x: auto;
	hyphens: none;
	@media (min-width: ${breakpoint}) {
		overflow: hidden;
		text-overflow: ellipsis;
	}
`;

const IndexLinks = () => {
	const { links, setLinksFromBrowser } = useLinks();
	const { copiedLink, copyLink } = useCopyLink();
	useEffect(setLinksFromBrowser, []);
	return (
		<TransitionGroup component={LinkContainer}>
			{links.map(({ _id, full }) => (
				<CSSTransition timeout={500} classNames="list-fade" key={_id} appear>
					<Link key={_id} className="link" data-testid="link">
						<FullLink>{full}</FullLink>
						<CopyShortLink shortUrl={_id} />
						<LinkCopyButton
							id={_id}
							copiedLink={copiedLink}
							copyLink={copyLink}
						/>
					</Link>
				</CSSTransition>
			))}
		</TransitionGroup>
	);
};

export default IndexLinks;
