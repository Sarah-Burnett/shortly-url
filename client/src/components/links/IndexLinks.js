import React, { useEffect } from 'react'
import styled from 'styled-components'
import CopyShortLink from './CopyShortLink'
import LinkCopyButton from './LinkCopyButton'
import { breakpoint, alignWidth, smallerWidth } from '../Styles'
import { useCopyLink } from '../../hooks/useCopyLink';
import { useLinks } from '../../hooks/useLinks';

const LinkContainer = styled.div`
    position: relative;
    top: -15vh;
    width: ${alignWidth};
    height: min-content;
    margin: auto;
    @media (min-width: ${breakpoint}) {
        width: ${smallerWidth};
        top: -10vh;
    }
`

const Link = styled.div`
    min-height: 15vh;
    background: #FFF;
    width: ${alignWidth};
    padding: 2vh 2vw;
    margin: 2vh 0 2vh 0;
    border-radius: 15px;
    @media (min-width: ${breakpoint}){
        display: flex;
        align-items: center;
        height: 10vh;
        width: ${smallerWidth}
    }
`

const FullLink = styled.div`
    flex: 3;
    height: 5vh;
    line-height: 5vh;
    width: 100%;
    overflow-y: auto;
    overflow-x: auto;
    hyphens: none;
    @media (min-width: ${breakpoint}) {
        overflow: hidden;
        text-overflow: ellipsis;
    }
`

 
const IndexLinks = () => {
    const { links, setLinksFromBrowser } = useLinks();
    const { copiedLink, copyLink } = useCopyLink();
    useEffect(setLinksFromBrowser, []);
    return( 
    <LinkContainer className="linkContainer">
        {
        links.map( ({_id, full}) => {
            return <Link key={_id} className="link">
                <FullLink>{full}</FullLink>
                <CopyShortLink shortUrl={_id}/>
                <LinkCopyButton id={_id} copiedLink={copiedLink} copyLink={copyLink}/>
            </Link> 
        })
        }
    </LinkContainer> 
    )
}

export default IndexLinks;