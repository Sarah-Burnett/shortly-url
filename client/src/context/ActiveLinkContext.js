import React, { useState, createContext } from 'react';

export const ActiveLinkContext = createContext();

export const ActiveLinkProvider = (props) => {
    const [ activeLink, setActiveLink] = useState({
        activeLink: [],
        process: false
    });

    return (
        <ActiveLinkContext.Provider value={[activeLink, setActiveLink]}>
            {props.children}
        </ActiveLinkContext.Provider>
    )
};