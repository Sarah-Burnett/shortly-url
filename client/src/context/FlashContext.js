import React, { useState, createContext } from 'react';

export const FlashContext = createContext();

export const FlashProvider = (props) => {
    const [ flash, setFlash ] = useState({
        flashMsg: "",
        isFlashVisible: false,
        isAutoHideFlash: true,
    });
    return (
        <FlashContext.Provider value={[flash, setFlash]}>
            {props.children}
        </FlashContext.Provider>
    )
};