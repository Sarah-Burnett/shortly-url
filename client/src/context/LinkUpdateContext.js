import React, { useState, createContext } from 'react';

export const LinkUpdateContext = createContext();

export const LinkUpdateProvider = (props) => {
    const [ linkUpdate, setLinkUpdate] = useState();
    return (
        <LinkUpdateContext.Provider value={[linkUpdate, setLinkUpdate]}>
            {props.children}
        </LinkUpdateContext.Provider>
    )
};