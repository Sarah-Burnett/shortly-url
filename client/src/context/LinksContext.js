import React, { useState, createContext } from 'react';

export const LinksContext = createContext();

export const LinksProvider = (props) => {
    const [ links, setLinks] = useState([]);

    return (
        <LinksContext.Provider value={[links, setLinks]}>
            {props.children}
        </LinksContext.Provider>
    )
};