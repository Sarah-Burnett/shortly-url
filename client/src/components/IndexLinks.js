import React, { useState, useEffect, useContext } from 'react'
import { LinksContext } from '../context/LinksContext';
import styled from 'styled-components'
import { FullLink } from './links/FullLink'
import { ShortLink } from './links/ShortLink'
import { LinkCopyButton } from './links/LinkCopyButton'
import { breakpoint } from '../Styles'

const LinkContainer = styled.div`
    position: relative;
    top: -15vh;
    width: 80vw;
    height: min-content;
    margin: auto;
    @media (min-width: ${breakpoint}) {
        top: -10vh;
    }
`

const Link = styled.div`
    min-height: 15vh;
    background: #FFF;
    width: 80vw;
    padding: 2vh 2vw;
    margin: 2vh 0 2vh 0;
    border-radius: 15px;
    @media (min-width: ${breakpoint}){
        display: flex;
        align-items: center;
        height: 10vh;
    }
`

export const Links = () => {
    const [ links, setLinks ] = useContext(LinksContext);
    const [ copiedLink, setCopiedLink ] = useState('');
    const copyToClipboard = (event, id) => {
        event.target.previousSibling.select()
        document.execCommand('copy')
        event.target.focus()
        setCopiedLink(id)
    };
    useEffect(() => {
        if (localStorage.shortlyLinks) setLinks(links => JSON.parse(localStorage.shortlyLinks))
    }, []);
    return( 
    <LinkContainer className="linkContainer">
        {
        links.map( ({_id, full}) => {
            return <Link key={_id} className="link">
                <FullLink fullUrl={full}/>
                <ShortLink shortUrl={_id}/>
                <LinkCopyButton id={_id} copyToClipboard={copyToClipboard} copiedLink={copiedLink}/>
            </Link> 
        })
        }
    </LinkContainer> 
    )
}

// <Link className="link" key={link[0]}><div className="fullLink">{link[1]}</div><ShortLink readOnly="readonly" className="shortLink" value={`https://rel.ink/${link[0]}`}><a href={`https://rel.ink/${link[0]}`}>{`https://rel.ink/${link[0]}`}</a></ShortLink><div className="button"><Button value={`https://rel.ink/${link[0]}`} onClick={copyToClipboard}>{copyButton}</Button></div></Link></LinkContainer>