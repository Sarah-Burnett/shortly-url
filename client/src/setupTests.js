import React from 'react';
import ReactDOM from 'react-dom';
import '@testing-library/jest-dom/extend-expect';
import { render, cleanup, fireEvent, screen, waitForElement } from '@testing-library/react';
import { rest } from 'msw';
import { setupServer } from 'msw/node';
import renderer from 'react-test-renderer';
import { UserProvider } from './context/UserContext';
import { LinksProvider } from './context/LinksContext';
import { ActiveLinkProvider } from './context/ActiveLinkContext';
import { ProcessLinkProvider } from './context/ProcessLinkContext';
import { FlashProvider } from './context/FlashContext';
import App from './App';

export const ComponentWrapper = (Component) => {
    return (
        <UserProvider>
            <LinksProvider>
                <ActiveLinkProvider>
                    <ProcessLinkProvider>
                        <FlashProvider>
                            <Component/>
                        </FlashProvider>
                    </ProcessLinkProvider>
                </ActiveLinkProvider>
            </LinksProvider>
        </UserProvider>
    )
}

export const rendersWithoutCrashing = (Component) => {
    const div = document.createElement('div');
    ReactDOM.render(<Component/>, div);
}

export const matchesSnapshot = (Component) => {
    const tree = renderer.create(<Component/>).toJSON();
    expect(tree).toMatchSnapshot();
}

export const sampleLinks = [
    {
        "date":"2020-08-14T17:29:53.041Z",
        "click":[
            {"date":1597426302926},
            {"date":1597426306071},
            {"date":1597426310231},
            {"date":1597426314763},
            {"date":1597428206995},
            {"date":1597428210093},
            {"date":1597428215131},
            {"date":1597428218755}
        ],
        "_id":"K1pjMCK-F",
        "full":"http://reactcommunity.org/react-transition-group/css-transition",
        "userId":"5ebe7d235e4a501ee8699058",
        "short":"octopus","__v":8
    },
    {
        "date":"2020-08-14T17:33:55.748Z",
        "click":[],
        "_id":"lsncCe5RV",
        "full":"https://ba-grant.netlify.app/",
        "userId":"5ebe7d235e4a501ee8699058",
        "short":"bagrant","__v":0
    }
];
