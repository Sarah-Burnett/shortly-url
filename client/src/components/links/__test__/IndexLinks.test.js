import React, { useState } from 'react'
import { rendersWithoutCrashing, matchesSnapshot, sampleLinks } from '../../../setupTests';
import IndexLinks from '../IndexLinks';
import { LinksContext } from '../../../context/LinksContext';
import { render, screen } from '@testing-library/react';

const IndexLinksWrapper = () => {
    const [ links, setLinks ] = useState(sampleLinks);
    return (
        <LinksContext.Provider value={[links, setLinks]}>
            <IndexLinks/>
        </LinksContext.Provider>
    )
};

it('renders index links without crashing', () => rendersWithoutCrashing(IndexLinksWrapper));

it('index links matches snapshot', () => matchesSnapshot(IndexLinksWrapper));

it('renders link for each link', () => {
    render(<IndexLinksWrapper/>);
    expect(screen.getAllByTestId('link').length).toBe(2);
});