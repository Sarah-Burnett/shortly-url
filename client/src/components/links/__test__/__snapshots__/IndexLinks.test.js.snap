// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`index links matches snapshot 1`] = `
<div
  className="sc-fznZeY gvBrXp linkContainer"
>
  <div
    className="sc-fzokOt jIsGrC link"
    data-testid="link"
  >
    <div
      className="sc-fzqBZW hSQcPn"
    >
      http://reactcommunity.org/react-transition-group/css-transition
    </div>
    <input
      className="sc-fzplWN dgYTEC shortLink"
      readOnly="readonly"
      value="https://short.ly/K1pjMCK-F"
    />
    <button
      className="sc-AxirZ sc-fznyAO dUXjER linkcopybtn"
      onClick={[Function]}
    >
      Copy
    </button>
  </div>
  <div
    className="sc-fzokOt jIsGrC link"
    data-testid="link"
  >
    <div
      className="sc-fzqBZW hSQcPn"
    >
      https://ba-grant.netlify.app/
    </div>
    <input
      className="sc-fzplWN dgYTEC shortLink"
      readOnly="readonly"
      value="https://short.ly/lsncCe5RV"
    />
    <button
      className="sc-AxirZ sc-fznyAO dUXjER linkcopybtn"
      onClick={[Function]}
    >
      Copy
    </button>
  </div>
</div>
`;
